File: Data Structures & Algorithms/max-water-container/submission-5.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        volume = 0
        l, r = 0, len(heights) - 1

        while l < r: 
            curr_vol = (r - l) * min(heights[l], heights[r])
            volume = max(curr_vol, volume)
            if heights[l] < heights[r]: 
                l += 1
            else: 
                r -= 1 

        return volume 

File: Data Structures & Algorithms/maximum-product-subarray/submission-5.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int curr_min = nums[0];
        int curr_max = nums[0]; 
        int res = nums[0]; 

        for (int i = 1; i < nums.size(); i++) { 
            int temp_min  = min({
                curr_min * nums[i], 
                curr_max * nums[i], 
                nums[i]
            });

            int temp_max = max({
                curr_min * nums[i], 
                curr_max * nums[i], 
                nums[i]
            });

            curr_min = temp_min;
            curr_max = temp_max; 

            res = max(res, curr_max); 
        }

        return res; 
    }
};
